CartSynchronizer: Do not emit changes until remote request is complete

sirbrillig opened the pull request:
When something in the UI causes a modification to the cart, the
`CartStore` creates a state transition function to update its state. It
then passes that transition function to the `CartSynchronizer` via its
`update()` method so that the state modification can be persisted to the
`shopping-cart` endpoint. Often the result of that request is not the
same as the modified local cart anyway, so this request is more than
just persistence. The `CartStore` submits this request, and then waits
for the `CartSynchronizer` to emit a change event before emitting its
own change event to its listeners. Typically these listeners are React
components that want to update their state to match the modified cart.

The `CartSynchronizer`'s `update()` method generates the updated local
state based on the transition function and then sends that updated state
as a request to the `shopping-cart` endpoint. When the request comes
back, the `CartSynchronizer` emits a change event.

However, the `CartSynchronizer` also emits a change event just after
it's submitted the endpoint request. This means that listener React
components will update their state based on possibly incorrect data that
will change when the endpoint request finishes. This optimistic update
can cause race conditions if the result of the update causes side
effects like other `shopping-cart` requests or UI actions that cannot
easily be undone when the new cart instance appears.

This change removes that second change event so that the
`CartSynchronizer`, and thus the `CartStore`, only emit a change when
the server has validated the data.

Fixes #40261 

#### Testing instructions

This is a tricky thing to test. Primarily, we just need to verify that shopping cart operations work as expected (adding and removing items from the cart) in the old checkout and other places in calypso (the new checkout uses a different mechanism), and hope that nothing is relying on this optimistic data.

sirbrillig committed:
CartSynchronizer: Do not emit changes until remote request is complete

When something in the UI causes a modification to the cart, the
`CartStore` creates a state transition function to update its state. It
then passes that transition function to the `CartSynchronizer` via its
`update()` method so that the state modification can be persisted to the
`shopping-cart` endpoint. Often the result of that request is not the
same as the modified local cart anyway, so this request is more than
just persistence. The `CartStore` submits this request, and then waits
for the `CartSynchronizer` to emit a change event before emitting its
own change event to its listeners. Typically these listeners are React
components that want to update their state to match the modified cart.

The `CartSynchronizer`'s `update()` method generates the updated local
state based on the transition function and then sends that updated state
as a request to the `shopping-cart` endpoint. When the request comes
back, the `CartSynchronizer` emits a change event.

However, the `CartSynchronizer` also emits a change event just after
it's submitted the endpoint request. This means that listener React
components will update their state based on possibly incorrect data that
will change when the endpoint request finishes. This optimistic update
can cause race conditions if the result of the update causes side
effects like other `shopping-cart` requests or UI actions that cannot
easily be undone when the new cart instance appears.

This change removes that second change event so that the
`CartSynchronizer`, and thus the `CartStore`, only emit a change when
the server has validated the data.

fditrapani commented:
I tested this in a couple cases and here's what I found:

- Added a domain to the cart, then remove it from the cart. Works as expected. It takes me back to the plans page and the cart icon clear. 
- Upgrade to a new plan from the plans page, then remove it from the cart. Returns to the plans page with a notification on the cart icon and the page is not scrollable. After a couple seconds, the notification goes away but I still can't scroll the page. I get some warnings in my console about "DevTools failed to parse SourceMap" but I don't know if this is related.
- Renewed a domain in redemption, then remove it from the cart. It's doing that visit to the plans page thing you're working on in another PR but then when I get to the cart, and remove the domain renewal, it behaves as expected.
- Upgrade a free site by adding a domain to the cart, removed the plan from the cart. Works as expected. It takes me back to the plans page and the cart icon clear. 
- Upgrade a free site by adding a domain to the cart, removed the domain first, then remove the plan from the cart. Returns to the plans page with a notification on the cart icon and the page is not scrollable. After a couple seconds, the notification goes away but I still can't scroll the page. I get some warnings in my console about "DevTools failed to parse SourceMap" but I don't know if this is related.

sirbrillig commented:
> It's doing that visit to the plans page thing you're working on in another PR

If you mean #40261, that's _this_ PR. So if you're still seeing it, that means that this doesn't solve the issue. 😭 I'll try to reproduce.

sirbrillig commented:
> Returns to the plans page with a notification on the cart icon and the page is not scrollable. After a couple seconds, the notification goes away but I still can't scroll the page.

This seems like a bigger issue. I created #40293 to track it.

sirbrillig commented:
> Renewed a domain in redemption, then remove it from the cart. It's doing that visit to the plans page thing you're working on in another PR but then when I get to the cart, and remove the domain renewal, it behaves as expected.

I can't reproduce this yet. I wonder if maybe this was specific to removing domain redemptions from the cart (disabled by #40296). One thing that could happen is that if the shopping cart on the server emptied the cart when you removed the redemption, then you'd end up on the plans page... but you said "when I get to the cart and remove the domain renewal" which implies the renewal was still there. 🤔 It would be really helpful to see the debug logs for this behavior.

sirbrillig commented:
I asked in Slack and @fditrapani clarified the above issue:

> I have a domain in redemption. When I pressed renew, it takes me to the plans page. From there, I can get to the cart and remove the renewal from it where everything behaves as expected.

So I think this is actually the issue that #40270 is trying to solve and is hopefully unrelated to this PR. 😌

fditrapani commented:
That's right! Sorry about the confusing messaging there.

sirbrillig commented:
I just rebased this on top of #40270 now that we (hopefully) have that one working. This will now need another rebase after that one is merged, but for now we can use this branch to test both together.

michaeldcain reviewed (approved):
This looks good. I'm a little nervous about unforeseen consequences because we don't fully know what depends on this, but I don't know of a better way to test it than we already have.

sirbrillig commented:
> I'm a little nervous about unforeseen consequences

Me too. 😓 But as you say, I don't know what more we could do. I'll drop a link for the team calypso folks to see if they could give it a quick look.

sirbrillig commented:
I'm pretty sure the failing e2e test "Select Business Plan" fails in master also. 

As for the "Can Remove Coupon" failing e2e test, I'm trying to reproduce that locally, since that does sound like it could be related to this change.

griffbrad commented:
I think the best we can do in this situation is to step through the dispatch-able actions on the store and ensure each one works as expected without the optimistically emitted pre-request event:
https://github.com/Automattic/wp-calypso/blob/master/client/lib/cart/store/index.js#L137
Reading through this thread so far, it seems like that's essentially what you've been doing, but perhaps glancing there could point out some additional potential areas of concern along with the current failing e2e.

sirbrillig committed:
e2e tests: Select plans segment for Business

Some of the e2e tests select the Business plan after visiting the Plans
tab, but the Business plan is not on the default plans segment, it is on
the second one. This change selects that segment before continuing.

sirbrillig committed:
e2e tests: Wait for coupon when applying

This waits for a coupon to be applied before continuing when running
`applyCoupon()`.

sirbrillig committed:
e2e tests: Wait for coupon when removing

This waits for a coupon to be removed before continuing when running
`removeCoupon()`.

sirbrillig commented:
The tests were failing partly because it looks like someone changed how the plans page is displayed, but more specific to this PR because the coupon tests were not waiting for the coupon round-trip to the server before checking the price. I pushed some changes to have it wait instead, as in practice the wait time is extremely short and I suspect the test code was just an oversight. I'll wait to see if any other tests are failing.

sirbrillig commented:
@griffbrad thanks! yes, all those actions should be fine!
